### ISSUE TITLE
Develop a robust, flexible walk implementation

### ISSUE THREAD

**Issue by jaraco** [jaraco/path#154]
In #152, @thombashi reports a failed expectation - that it should be possible not to follow symlinks when walking a directory tree. Investigating, we learn that not following symlinks is the default behavior for `os.walk`, which may be a good model to follow.

But in exploring the prospect, I wonder if there's perhaps an even more simple yet extensible design, something in which the caller can more generally customize the filtering and traversal behavior of the walk, something like:

```
def walk(self, include, traverse, errors):
    ...
```

I'd also like to explore the possibility of using yield expressions to possibly eliminate the need for a traverse parameter.

**Comment by jaraco** [comment]
Today, I explored the possibility of using yield from and send calls to alter the traversal behavior on demand:

```
path.py master $ def yield_send_example():
................     print((yield 1))
................     print((yield 2))
................     print((yield 3))
................
path.py master $ gen = yield_send_example()
path.py master $ gen.send('a')
TypeError: can't send non-None value to a just-started generator
path.py master $ next(gen)
1
path.py master $ gen.send('b')
b
2
path.py master $ next(gen)
None
3
```

It seems to me we may be able to allow a caller (or wrapper) to 'send' special values back to the generator to suppress or extend traversal on specific directories (such as symlinks).

I'll put together a demo of this concept later.

**Comment by jaraco** [comment]
In the [feature/traversal-control branch](/jaraco/path.py/tree/feature/traversal-control), I now have a working draft of such a feature. It demonstrates how a generator like `Path.walk` can defer to the result of the `yield expression` to determine the traversal behavior, and then supplies a `Traversal` class to facilitate the translation of a simple `follow` function to apply to each child as it's encountered.

**Comment by jaraco** [comment]
Fixed in #193 for release with v14.0.0.
